### PR TITLE
planner: fix wrong outer join elimination when join key is empty

### LINKED ISSUE #15846: Incorrect result for LEFT JOIN and NULL values
Consider the following statements:
```sql
CREATE TABLE t0(t0 INT UNIQUE);
CREATE TABLE t1(c0 FLOAT);
INSERT INTO t1(c0) VALUES (0);
INSERT INTO t0(t0) VALUES (NULL), (NULL);
SELECT t1.c0 FROM t1 LEFT JOIN t0 ON 1; -- expected: {0, 0}, actual: {0}
```
I would expect that the `SELECT` fetches two rows, but only one is fetched. When removing the `UNIQUE` from `t0`, the query works correctly:

```sql
CREATE TABLE t0(t0 INT);
CREATE TABLE t1(c0 FLOAT);
INSERT INTO t1(c0) VALUES (0);
INSERT INTO t0(t0) VALUES (NULL), (NULL);
SELECT t1.c0 FROM t1 LEFT JOIN t0 ON 1; -- {0, 0}
```
Environment:
```
mysql> SELECT tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                               |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-75-ga6de0e38d-dirty
Git Commit Hash: a6de0e38d49c97671d316590c8c945eb518ca2b2
Git Branch: master
UTC Build Time: 2020-03-26 12:11:33
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #15846  

Problem Summary: Wrong outer join elimination leads to wrong result.

### What is changed and how it works?

What's Changed: Check join key num in function `isInnerJoinKeysContainIndex` (Empty join key doesn't contains index.)

How it Works: `CREATE TABLE t0(t0 INT UNIQUE);` will create an index in t0. So the possibleAccessPaths will be two. But the second check causes the wrong result.
The path.IdxCols  is empty.

### Related changes

- Need to cherry-pick to the release branch: 4.0

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 
Fix wrong outer join elimination when join key is empty.
(If need)

### CODE REVIEW

File: planner/core/rule_join_elimination.go [hunk 1]
```diff
@@ -122,7 +122,7 @@ func (o *outerJoinEliminator) isInnerJoinKeysContainUniqueKey(innerPlan LogicalP
 // check whether one of index sets is contained by inner join index
 func (o *outerJoinEliminator) isInnerJoinKeysContainIndex(innerPlan LogicalPlan, joinKeys *expression.Schema) (bool, error) {
 	ds, ok := innerPlan.(*DataSource)
-	if !ok {
+	if !ok || joinKeys.Len() == 0 {
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go]: Should we check the `joinKeys.Len() == 0 && len(otherConditions) > 0`?

Comment (wshwsh12) [reply planner/core/rule_join_elimination.go]: Why we need to check `otherConditions`? In this case, len(otherConditions) == 0.

File: planner/core/rule_join_elimination.go [hunk 2]
```diff
@@ -132,6 +132,9 @@ func (o *outerJoinEliminator) isInnerJoinKeysContainIndex(innerPlan LogicalPlan,
 		if !path.Index.Unique {
 			continue
 		}
+		if path.IdxCols == nil || len(path.IdxCols) == 0 {
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go:135]:
```suggestion
		if len(path.IdxCols) == 0 {
```

Comment (lzmhhh123) [reply planner/core/rule_join_elimination.go:135]: And we can merge the above condition and this one.

**Comment by XuHuaiyu** [comment]
Do we need to cherry-pick this to 3.1 and 3.0?

**Comment by wshwsh12** [comment]
> Do we need to cherry-pick this to 3.1 and 3.0?

No.

**Pull request merged**